app/Filament/Resources/UserResource/Pages/ListUsers.php

```diff
@@ -167,4 +178,39 @@ protected function getHeaderActions(): array
             CreateAction::make(),
         ];
     }
+
+    protected function getRolesOption(): array
+    {
+        $roles = Role::query()->take(50)->pluck('name', 'id')->toArray();
+
+        return [
+            '' => [
+                'none' => 'None',
+            ],
+            'Roles' => [
+                ...$roles,
+            ],
+        ];
+    }
+
+    protected function roleFilter(Builder $query, array $data): void
+    {
+        if (empty($data['values'])) {
+            return;
+        }
+
+        $query->where(function ($query) use ($data) {
+            $filteredValues = $data['values'];
+            $query->when(in_array('none', $filteredValues), function ($query) {
+                $query->whereDoesntHave('roles');
+            })
+                ->orWhereHas('roles', function ($query) use ($filteredValues) {
+                    if (in_array('none', $filteredValues)) {
+                        unset($filteredValues[array_search('none', $filteredValues)]);
+                    }
```

Comment:
Instead of unsetting, could you maybe just do an `array_filter()` which gets passed directly to the `whereIn()`?